sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcUtils.scala

```diff
@@ -0,0 +1,288 @@
+    if (value == null) {
+      null
+    } else {
+      dataType match {
+        case NullType => null
+
+        case BooleanType => new BooleanWritable(value.asInstanceOf[Boolean])
+
+        case ByteType => new ByteWritable(value.asInstanceOf[Byte])
+        case ShortType => new ShortWritable(value.asInstanceOf[Short])
+        case IntegerType => new IntWritable(value.asInstanceOf[Int])
+        case LongType => new LongWritable(value.asInstanceOf[Long])
+
+        case FloatType => new FloatWritable(value.asInstanceOf[Float])
+        case DoubleType => new DoubleWritable(value.asInstanceOf[Double])
+
+        case StringType => new Text(value.asInstanceOf[UTF8String].getBytes)
+
+        case BinaryType => new BytesWritable(value.asInstanceOf[Array[Byte]])
+
+        case DateType => new DateWritable(DateTimeUtils.toJavaDate(value.asInstanceOf[Int]))
+
+        case TimestampType =>
+          val us = value.asInstanceOf[Long]
+          var seconds = us / DateTimeUtils.MICROS_PER_SECOND
+          var micros = us % DateTimeUtils.MICROS_PER_SECOND
+          if (micros < 0) {
+            micros += DateTimeUtils.MICROS_PER_SECOND
+            seconds -= 1
+          }
+          val t = new OrcTimestamp(seconds * 1000)
+          t.setNanos(micros.toInt * 1000)
+          t
+
+        case _: DecimalType =>
+          new HiveDecimalWritable(HiveDecimal.create(value.asInstanceOf[Decimal].toJavaBigDecimal))
+
+        case st: StructType =>
+          convertInternalRowToOrcStruct(value.asInstanceOf[InternalRow], st)
+
+        case ArrayType(et, _) =>
+          val data = value.asInstanceOf[ArrayData]
+          val list = createOrcValue(dataType)
+          for (i <- 0 until data.numElements()) {
+            list.asInstanceOf[OrcList[WritableComparable[_]]]
+              .add(getWritable(data.get(i, et), et))
+          }
+          list
+
+        case MapType(keyType, valueType, _) =>
+          val data = value.asInstanceOf[MapData]
+          val map = createOrcValue(dataType)
+            .asInstanceOf[OrcMap[WritableComparable[_], WritableComparable[_]]]
+          data.foreach(keyType, valueType, { case (k, v) =>
+            map.put(
+              getWritable(k.asInstanceOf[Object], keyType),
+              getWritable(v.asInstanceOf[Object], valueType))
+          })
+          map
+
+        case udt: UserDefinedType[_] =>
+          val udtRow = new SpecificInternalRow(Seq(udt.sqlType))
+          udtRow(0) = value
+          convertInternalRowToOrcStruct(udtRow,
+            StructType(Seq(StructField("tmp", udt.sqlType)))).getFieldValue(0)
+
+        case _ =>
+          throw new UnsupportedOperationException(s"$dataType is not supported yet.")
+      }
+    }
+
+  }
+
+  /**
+   * Return Spark Catalyst value from WritableComparable object.
+   */
+  private[orc] def getCatalystValue(value: WritableComparable[_], dataType: DataType): Any = {
```

Comment:
we'd better return a function to avoid per-row pattern matche. cc @HyukjinKwon who fixed similar problems many times.